richtextfx/src/main/java/org/fxmisc/richtext/model/ReadOnlyStyledDocument.java

```diff
@@ -67,143 +66,124 @@ public Summary reduce(Summary left, Summary right) {
     private static final BiFunction<Summary, Integer, Either<Integer, Integer>> NAVIGATE =
             (s, i) -> i <= s.length() ? left(i) : right(i - (s.length() + 1));
 
-    public static <PS, S> ReadOnlyStyledDocument<PS, S> fromString(String str, PS paragraphStyle, S style) {
+    public static <PS, SEG, S> ReadOnlyStyledDocument<PS, SEG, S> fromString(String str, PS paragraphStyle, S style, SegmentOps<SEG, S> segmentOps) {
         Matcher m = LINE_TERMINATOR.matcher(str);
 
         int n = 1;
         while(m.find()) ++n;
-        List<Paragraph<PS, S>> res = new ArrayList<>(n);
+        List<Paragraph<PS, SEG, S>> res = new ArrayList<>(n);
 
         int start = 0;
         m.reset();
         while(m.find()) {
             String s = str.substring(start, m.start());
-            res.add(new Paragraph<>(paragraphStyle, s, style));
+            res.add(new Paragraph<>(paragraphStyle, segmentOps, s, style));
             start = m.end();
         }
         String last = str.substring(start);
-        res.add(new Paragraph<>(paragraphStyle, last, style));
+        res.add(new Paragraph<>(paragraphStyle, segmentOps, last, style));
 
-        return new ReadOnlyStyledDocument<>(res);
+        return new ReadOnlyStyledDocument<>(res, segmentOps);
     }
 
-    public static <PS, S> ReadOnlyStyledDocument<PS, S> from(StyledDocument<PS, S> doc) {
+    public static <PS, SEG, S> ReadOnlyStyledDocument<PS, SEG, S> from(StyledDocument<PS, SEG, S> doc) {
         if(doc instanceof ReadOnlyStyledDocument) {
-            return (ReadOnlyStyledDocument<PS, S>) doc;
+            return (ReadOnlyStyledDocument<PS, SEG, S>) doc;
         } else {
-            return new ReadOnlyStyledDocument<>(doc.getParagraphs());
+            return new ReadOnlyStyledDocument<>(doc.getParagraphs(), doc.getSegOps());
         }
     }
 
-    /**
-     * Creates a new ReadOnlyStyledDocument with one Segment.
-     * The resulting ReadOnlyStyledDocument can be inserted or appended to a StyledTextArea.
-     *
-     * @param seg   The segment which shall be contained in the document.
-     * @param paragraphStyle The paragraph style to use for the paragraph which contains the segment.
-     *
-     * @return A ReadOnlyStyledDocument with the given segment.
-     */
-    public static <PS, S> ReadOnlyStyledDocument<PS, S> from(Segment<S> seg, PS paragraphStyle) {
-        List<Paragraph<PS, S>> res = new ArrayList<>(1);
-        Paragraph<PS, S> content = new Paragraph<PS, S>(paragraphStyle, Arrays.asList(seg));
-        res.add(content);
-        return new ReadOnlyStyledDocument<>(res);
-    }
```

Comment:
I am not completely sure which diff that is - this should be the method which creates a paragraph from an arbitrary segment, it might have been removed when merging from the genericSegment branch, but it has been re-added in https://github.com/TomasMikula/RichTextFX/pull/356/commits/ef989df7fd38d24293cd9f1efa7dbb79727a8bf3


```
public static <PS, SEG, S> ReadOnlyStyledDocument<PS, SEG, S> fromSegment(SEG segment,  PS paragraphStyle, S style, SegmentOps<SEG, S> segmentOps) {
    List<Paragraph<PS, SEG, S>> res = new ArrayList<>(1);
    res.add(new Paragraph<>(paragraphStyle, segmentOps, segment));
    return new ReadOnlyStyledDocument<>(res, segmentOps);
}
```

I have adjusted it to the original code using Arrays.asList(), probably thats even better ....